[vulkan] Enable prepacked addmm for linear layers

SS-JIA opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47815 [vulkan] Enable prepacked addmm for linear layers**

Differential Revision: [D24908605](https://our.internmc.facebook.com/intern/diff/D24908605)

Commit:
[vulkan] Enable prepacked addmm/mm for linear layers

[ghstack-poisoned]

Commit:
Update on "[vulkan] Enable prepacked addmm/mm for linear layers"




[ghstack-poisoned]

IvanKobzarev commented on torch/csrc/jit/passes/vulkan_rewrite.cpp:
```diff
@@ -22,6 +22,54 @@ namespace jit {
+    const auto& match_vmap = match.values_map;
+    auto linear_value = match_vmap.at(vmap.at("linear"));
+    auto func_name = graph_rewrite_helper::getFuncName(linear_value);
+    if (func_name == "linear") {
+      return true;
+    }
+    return false;
```
return (func_name == "linear");

IvanKobzarev commented on aten/src/ATen/native/vulkan/ops/Mm.cpp:
```diff
@@ -163,6 +307,37 @@ Tensor mm(const Tensor& self_arg, const Tensor& mat2_arg) {
 
 #ifdef USE_VULKAN_API
 
+TORCH_LIBRARY(vulkan_lin, m) {
```
Can we reuse the same library and namespace as for conv 'vulkan_prepack' ?

SS-JIA replied:
This can be done, but just as a heads up I'll have to do all the registrations in a new file (like `RegisterOpContextClass.cpp` for XNNPACK.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/ops/Mm.cpp:
```diff
@@ -7,94 +8,207 @@ namespace vulkan {
+vTensor pack_weights(api::Resource::Pool& pool, const Tensor& weight_arg) {
+  if (weight_arg.is_vulkan()) {
+    return convert(weight_arg);
+  }
```
This check is no longer needed.  weight_arg.vulkan() will not be an expensive call if weight_arg.is_vulkan() is true so you can just replace both code paths with `return convert(weight_arg.vulkan());`

AshkanAliabadi commented on aten/src/ATen/native/vulkan/ops/Mm.cpp:
```diff
@@ -7,94 +8,207 @@ namespace vulkan {
-         (self_sizes[Layout::Parameter::height] ==
-            mat2_sizes[Layout::Parameter::width])),
-        "Incompatible matrix dimensions!");
+  if (bias_arg) {
```
You can replace both ifs with the following:

```
if (bias_arg) {
  return convert(bias_arg->vulkan());
}

// The else logic goes here.  No need for else since you are returning from if.
```

AshkanAliabadi commented on aten/src/ATen/native/vulkan/ops/Mm.cpp:
```diff
@@ -7,94 +8,207 @@ namespace vulkan {
+  bool valid =
+      (input.sizes()[Layout::Parameter::width] ==
+       weight.sizes()[Layout::Parameter::height]);
+  return valid;
```
```
return input.sizes()[Layout::Parameter::width] ==
           weight.sizes()[Layout::Parameter::height];
```

AshkanAliabadi commented on torch/csrc/jit/passes/vulkan_rewrite.cpp:
```diff
@@ -22,6 +22,54 @@ namespace jit {
+        %res = vulkan_prepack_lin::linear_run(%input, %packed_weight_bias)
+        return (%res))";
+
+  auto filter = [](const Match& match,
```
const

AshkanAliabadi commented on torch/csrc/jit/passes/vulkan_rewrite.cpp:
```diff
@@ -22,6 +22,54 @@ namespace jit {
+  auto filter = [](const Match& match,
+                   const std::unordered_map<std::string, Value*>& vmap) {
+    const auto& match_vmap = match.values_map;
+    auto linear_value = match_vmap.at(vmap.at("linear"));
```
const please

AshkanAliabadi reviewed (approved):
Excellent, thanks!

Commit:
Update on "[vulkan] Enable prepacked addmm for linear layers"


Differential Revision: [D24908605](https://our.internmc.facebook.com/intern/diff/D24908605)

[ghstack-poisoned]

Commit:
Update on "[vulkan] Enable prepacked addmm for linear layers"


Differential Revision: [D24908605](https://our.internmc.facebook.com/intern/diff/D24908605)

[ghstack-poisoned]
